Make NestedTensor compilable with eager backend

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110529
* __->__ #109171

In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of @bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

soulitzer committed:
Make NestedTensor compilable with eager backend

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2378,7 +2383,7 @@ def create_symbolic_sizes_strides_storage_offset(
         def maybe_specialize_sym_int_with_hint(maybe_sym) -> int:
             assert isinstance(maybe_sym, (int, torch.SymInt))
-            if isinstance(maybe_sym, SymInt):
+            if is_symbolic(maybe_sym):
```
you lost the int-ness test here

soulitzer replied:
We assert the int-ness above :)

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2581,7 +2586,11 @@ def create_symbol(
+            if isinstance(val, int):
+                return sympy.Integer(val)
+            else:
+                return sympy.Symbol(str(val))
```
this is hella suspicious. Why?

soulitzer replied:
Creating an Integer requires passing in a plain int, which we don't have.

ezyang replied:
I think it would probably be better not to try to inject singletons into sympy's symbolic expression universe, esp since we don't really support any nontrivial arithmetic on them (except for comparisons against 0/1, but I'm guessing this can be special cased). One reason I'm not keen on treating singletons as sympy Symbols is that symbols don't have correct semantics for singletons:

```
>>> import sympy
>>> from sympy.abc import j0, j1
>>> sympy.Eq(j0, j1)
Eq(j0, j1)
```

This is wrong: the singletons j0 and j1 can NEVER equal each other. But sympy doesn't know that, because to sympy, these are not skolem variables; they could get substituted later.

soulitzer replied:
As discussed offline to represent singleton int as a subclass of sympy.AtomExpr instead.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2822,6 +2835,8 @@ def get_symbol(tensor_dim_src):
         # tensors that never actually become graph arguments (they are
         # pruned).  In this case, only Dynamo knows about these arguments.
         def track_symint(source, val, constraint=None):
+            assert not isinstance(val, SymInt) or is_symbolic(val)
```
I'm confused, because we immediately turn around and test `isinstance(val, SymInt)` below

soulitzer replied:
Oh this was just testing that we don't expect to see any singleton symint, but yes if this function only takes symint, the first part of the check is redundant.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2900,22 +2915,29 @@ def hint(s):
+                inner_sources_and_tensors = [(AttrSource(source, attr), getattr(t, attr)) for attr in attrs]
+                if t.is_nested:
+                    # For NestedTensors we also need to track the ragged symint
+                    # on the outer tensor
```
@bdhirsh could you please take a look at this

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1631,7 +1634,7 @@ def wrap_to_fake_tensor_and_record(
         tx.output.tensor_weakref_to_sizes_strides[WeakIdRef(e)] = {
             "size": fake_e.size(),
-            "stride": fake_e.stride(),
+            "stride": fake_e.stride() if not e.is_nested else None,
```
support strides on jagged tensor, you know you want toooooo

soulitzer replied:
Ahh I've really got to think about what that means more lol

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -442,6 +445,24 @@ def empty_create(inner_t, inner_src):
+                                ),
+                                hint=r._size[1],
+                            )
+                            r.set_raggedness_id(sym_ragged_size)
```
What's this?

soulitzer replied:
With the current fakification logic
- (1) we fakify the inner tensors and pass them to `__torch_unflatten__`
- (2) user is responsible for reconstruct the outer subclass, e.g. construct the sizes of outer tensor using whatever symints are available from the inner tensors

Sometimes it is possible to derive the outer sizes from the inner ones: for example, I can compute the symint for the `B` in `[B, *, D]` if I do `offsets.shape[0] - 1`.

In the case of the ragged dimension, this is not possible, however, and so we need to do something special.

There are probably a few options here, adding this `set_raggedness_id` method to smash the symbolic raggedness dimension onto the wrapper subclass is one of them. Alternatively, we could've extended __tensor_unflatten__ to accept an additional `source` argument, and constructed the raggedness symint from inside there.

I don't have a strong opinion on which option which pick here, but the idea was to avoid modifying the common `__tensor_unflatten__` API given that we don't have any other examples of subclasses to go off of.

ezyang commented:
So, there is a pending design discussion we haven't had yet, which is how are you handling "dynamic" ragged dimensions. I'm not sure what your strategy is. You seem to be YOLO creating a sympy Symbol. This seems... unlikely. Can you elaborate in more detail?

soulitzer commented:
At a high level, the only difference between sympy.Integer and sympy.Symbol is that Integer is just a Symbol with some assumptions baked so that simplification can be done. Since we don't allow any arithmetic on our Symbol, I didn't expect much simplification to be possible with it anyway.

ezyang commented:
Let me help you with some of the implications.

Because singletons are only equal to themselves, you need different reasoning principles for them. In particular, it may be most straightforward to simply assume that the aliasing pattern must match exactly from run to run. For example, suppose I have:

f(x: NT[64, j0, 8], y: NT[64, j0, 8]): ...

Here, the fact that f has the same j0 for x and y means that their jagged dimensions are equal and I can, e.g., add these two JTs together without needing padding/truncation. If in the future I call f with `x, y: NT[64, j1, 8]`, this should work, but if I call it with `x: NT[64, j0, 8], y: NT[64, j1, 8]`, this should NOT work, because I have compiled code assuming that the jagged dimensions are equal.

So the guard structure has to be much more similar to the aliasing guard structure @voznesenskym added for tensors: you need to ensure equivalent singletons go to the same symbol (let's call these singleton symbols) and generate equality guards for them (same as regular symbols). I'm not sure if you need to generate distinctness guards, this depends on whether or not generated code for j0, j1 would work if the jagged dimensions actually are equal e.g., j0, j0.

You're going to need a design doc for this.

cc @nunoplopes

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh commented on torch/_dynamo/guards.py:
```diff
@@ -1078,7 +1084,7 @@ def convert(size_or_stride):
                 requires_grad = t.requires_grad
                 sizes = dynamic_dims_sizes[i]
-                strides = dynamic_dims_strides[i]
+                strides = dynamic_dims_strides[i] if not t.is_nested else "None"
```
Maybe worth a comment on what the dynamic shapes strides strategy is for nested?

bdhirsh commented on torch/_subclasses/meta_utils.py:
```diff
@@ -406,9 +406,12 @@ def is_c_of_r(complex_dtype, real_dtype):
+                            sizes,
+                            strides,
+                            storage_offset,
+                        ) = sym_sizes_strides_storage_offset(t, source)
```
Hmmm, what do you think of rearranging so that `def sym_sizes_strides_storage_offset(x, x_src)` checks if `x` is nested, and does the special handling in there (and maybe calls `NestedTensor.create_raggedness_id()` or something).

One reason this is annoying to reason about is because you are now not creating `sizes/strides` at all at the top-level, but there's a bunch of code (line 492 and below) that expects `sizes/strides` to exist, so it can detect happy-path code (like when we're fakeifying a tensor with 0 storage_offset).

Another thing is that doing the NestedTensor logic inside of that function should more naturally support composing subclasses (say we have `TwoTensor(NestedTensor, NestedTensor)` - it would be nice if inner tensors that are nested get fakefied the right way automatically).

soulitzer replied:
Good point, I didn't think about what would happen when subclasses compose. Updating the logic to happen inside `sym_sizes_strides_storage_offset`.

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented:
In the latest update we continue to represent singleton Symbolically within the framework of sympy. Unlike the previous iteration however, we use a subclass of AtomicExpr instead of Symbol. 

`sympy.Symbol` had two main drawbacks:
- the semantics do not match that of Singleton ints as pointed out in [comment](https://github.com/pytorch/pytorch/pull/109171#discussion_r1326158609)
- after an expression has been subbed with var_to_val, its free symbol count should be 0, but using sympy.Symbol would violate that assumption

Furthermore, we can subclass Atom (probably Symbol previously as well) to alter its semantics to match that of Singleton int.

```python
class _SingletonInt(sympy.AtomicExpr):
    def __init__(self, name):
        self.name = name
        super().__init__()

    def _eval_Eq(self, other):
        # TODO: support more cases
        if isinstance(other, _SingletonInt) and other.name == self.name:
            return sympy.true
        return sympy.false

    # This is necessary so that calling expr.free_symbols on exprs that contain
    # this Singleton does not error
    @property
    def free_symbols(self):
        return set()

print(sympy.Eq(Singleton("j0"), Singleton("j0")))  # True
print(sympy.Eq(Singleton("j0"), Singleton("j1")))  # False
print(Singleton("j0").free_symbols)  # set()
```

I'm still thinking about the following:
- consequences of actually supporting strides with singleton int
- code organization in meta utils to support nesting of subclasses (partially depends on the previous item)

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Since we disallow calling strides on NestedTensor (since we don't consider as "strided"), we add if-else's in the code to skip the parts that call strides. We do the same for storage offset, though maybe we can support that if we'd like. For now it is convenient enough to skip both.
- Verbose guard logic now handles ConstantSymBool SymBools
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We some one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.
SymIntify guards.cpp so that SingletonSymInt can be passed around there.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -470,8 +480,11 @@ def empty_create(inner_t, inner_src):
-                        and r.stride() == strides
-                        and r.storage_offset() == storage_offset
+                        and r.is_nested
+                        or (
```
worried about operator precedence here, can you explicitly parenthesize

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -441,6 +450,7 @@ def empty_create(inner_t, inner_src):
                                     AttrSource(source, attr),
                                 )
                             ),
+                            extra_context=extra_context,
```
I don't understand what's going on here, ISTR we talked about this a little but I've forgotten, can you remind me?

bdhirsh replied:
The problem is that:

* `transform_subclass()` is supposed to take a wrapper subclass, and produce a new wrapper subclass
* It also takes in a lambda, telling it how each inner tensor should be transformed.
* This isn't enough though - `transform_subclass` assumes that the **only** thing that changes from the input wrapper subclass to the output wrapper subclass is the values of the inner tensors (e.g. real -> fake). But for JaggedTensor, the extra `SymInt` corresponding to the jagged size needs to come from somewhere, and doesn't exist on the input jagged subclass (well... I think it does, but it doesn't have a proper `Source`?)


I also want to confirm my understanding: the jagged size is an actual `SymNode` in **both** eager and in torch.compile, but we can't just re-use the input `JaggedTensor`'s `SymNode` when we fakeify, because the eager mode jagged size doesn't have a source attached to it?

soulitzer replied:
> I also want to confirm my understanding: the jagged size is an actual SymNode in both eager and in torch.compile, but we can't just re-use the input JaggedTensor's SymNode when we fakeify, because the eager mode jagged size doesn't have a source attached to it? Since sources are specific to dynamo.

Hmm it depends on what you by "actual SymNode". Jagged tensor's ragged size is a SymInt holding a SymNode, but not actually the SymNode defined in Python! There's a pybinded SymNode class previously only used for XLA that we're coopting (see `torch/csrc/jit/python/init.cpp`). That is also why when we fakify we need to create a real python "class SymNode" using the pybinded SymNode as val.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -96,6 +97,41 @@ def uninteresting_files():
+class _SingletonInt(sympy.AtomicExpr):
+    def __init__(self, val, factor):
+        self.val = val
+        self.factor = factor
```
what's a factor?

soulitzer replied:
oh coefficient is probably a better name, I've updated "factor" -> "coeff" everywhere

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2993,8 +3040,9 @@ def create_symbol(
                 self.var_to_range[sympy_expr] &= constraint_dim.vr
```
OK, I see you didn't modify the var_to_range code, it doesn't really make sense to have ranges for singletons, not sure if you stub things out there, is it possible to remove the >= tests? It would be nice if this was the case 🤔

soulitzer replied:
I'm not sure if this actually addreses your question, but I recently update singleton ints to error (instead of returning False) when doing inequalities: >=, etc. (it's the first PR in this stack) for an (un?)related reason. Some small exceptions are made however , e.g.  j0 >= c returns True when c is <= 2.

ezyang replied:
I guess my question is, are you preventing var range tests on singleton ints? If so, we should avoid setting `var_to_range` for these.

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1707,9 +1708,11 @@ def wrap_to_fake_tensor_and_record(
+        if not e.is_nested:
+            dynamic_dims, constraint_dims = _automatic_dynamic(
+                e, tx, source.name(), static_shapes
+            )
```
I think... we do want automatic dynamic to run on nested tensor.

ezyang replied:
a TODO is probably ok for now

ezyang commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -20,12 +20,22 @@ def get_tensor_id(tensor, factor):
     _offsets: torch.Tensor
-    _size: Tuple[int, int, int]
+    _size: Tuple[int, torch.SymInt, int]
+    ragged_size: torch.SymInt
```
doc here would be good. Why isn't this underscored?

soulitzer replied:
Underscored, doc coming soon.

ezyang commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -20,12 +20,22 @@ def get_tensor_id(tensor, factor):
+        *,
+        sym_size=None,
+        sym_stride=None,
+        ragged_size=None,
```
You don't actually use this?

soulitzer replied:
Now I do!

ezyang commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -83,6 +102,45 @@ def __repr__(self):
+                    TensorPropertySource(source, TensorProperty.SIZE, 1),
+                ),
+                hint=ragged_size,
+            )
```
@bdhirsh can you review this?

But this looks suspicious

bdhirsh replied:
I think `__tensor_unflatten__()` is not the right place to allocate the extra symint.

In the steady-state, we don't want this to allocate a fresh symint on every round trip of flatten + unflatten:
```
attrs, meta = t.__tensor_flatten__()
new_t = type(t).__tensor_unflatten([getattr(t, x) for x in attrs], meta)
```

So... when we call `__tensor_unflatten__(inner_ts, meta)`, we probably want `meta` to already have the extra needed symint, so `__tensor_unflatten__` can just do the dummy "create an instance of the subclass and dump all of the internal tensors/metadata into it" without having to allocate anything.

bdhirsh replied:
Two options that seem (hopefully) reasonable to me are:

(1) Just special case `meta_utils.py` to do the extra symint allocation directly and have that be the extra context

(2) Figure out a more general mechanism for meta_utils.py to query the subclass and ask it do to extra work (like constructing an extra symint).

If we're already stuck with hardcoding `is_nested` in the fakeifying logic, then (1) seems ok

soulitzer replied:
Makes sense! Updating to go with (1）

soulitzer replied:
Just out of curiousity, what are situations where we might have multiple round trips of flatten + unflatten? My current understanding is that I go into meta_utils to fakify once, and then when I do operations, I am responsible for doing unwrap/rewrap in my torch dispatch function.

bdhirsh replied:
We go through it once to initially fakeify in dynamo, yep.

And we'll go through it again at least once in AOTAutograd: during functionalization, AOTAutograd will replace every input `SubclassTensor(FakeTensor)` with a `SubclassTensor(FunctionalTensor(FakeTensor))`: https://github.com/pytorch/pytorch/pull/104483/files#diff-df954bbf954d2dcb81f687876053267ffa4ddb36ed86b7d2bd76319ff2b94416R469.

We also unflatten at runtime in AOTAutograd, when taking the "plain tensor" outputs of the compiled graph, and wrapping them back up into subclasses to return to the user (if the user's captured graph expects outputs to be tensor subclasses).

ezyang commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -51,20 +61,29 @@ def __new__(cls, values, offsets, **kwargs):
+            B = offsets.shape[0] - 1
+            # TODO: factor out and generalize the stride computing logic
+            self._size = (B, self.ragged_size, D)
+            self._strides = (self.ragged_size * D, D, 1)
```
What's going on with the conditional here?

jbschlosser replied:
yeah I'm unclear on this too. Why do we need a constructor branch that directly accepts `sym_size` / `sym_stride`? Given the same `offsets`, we should produce the same `_size`, `_strides`, `ragged_size`, no?

soulitzer replied:
The `else` path here is for during eager, where I want to be able to pass in a offsets tensors and have the  ragged_size  be produced automatically with `get_tensor_id`

During tracing, that is not enough because `get_tensor_id` would return us the non-symbolic size.

I can see an alternative where we have a map somewhere that remembers what offsets map to what SymInts, but I don't have a preference.

soulitzer replied:
Actually, I found a better way to do this.

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -83,6 +102,45 @@ def __repr__(self):
+    def __tensor_unflatten__(inner_tensors: Dict, meta):
+        assert len(inner_tensors) == 2
+        values = inner_tensors["_values"]
+        offsets = inner_tensors["_offsets"]
```
do we get the exact same `offsets` and `values` tensors after a cycle of `__tensor_flatten__()` -> `__tensor_unflatten__()` ?

soulitzer replied:
It would depend on what kind of function you pass to `transform_subclass` (which is usually when the flatten/unflatten cycle happens). During fakification, for example, the fakification callback is passed, and so the we would get the fakified versions of offsets and values back.

jbschlosser reviewed (commented):
I checked out the subclass-related section of this PR and have a few questions as well

bdhirsh commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -83,6 +102,45 @@ def __repr__(self):
+        (requires_grad, ragged_size, *extra_meta) = meta
+
+        if len(extra_meta) > 0:
+            (source,) = extra_meta
```
commented below (but I don't think we can do this - if we're going to plumb extra metadata, we'll have to directly plumb the symint)

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- Adds a sympy `AtomicExpr` subclass to represent the semantics of singleton ints. Similar to how ints need to be converted to sympy.Integer, singleton ints would then be converted to this subclass.
- Update `isinstance(val, torch.SymInt)` to `is_symbolic` in various places to account for the existence of Singleton SymInts
- NestedTensor skips automatic dynamic
- SymIntify guards.cpp so that SingletonSymInt can be passed around there
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (going to write this up more)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- Adds a sympy `AtomicExpr` subclass to represent the semantics of singleton ints. Similar to how ints need to be converted to sympy.Integer, singleton ints would then be converted to this subclass.
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (going to write this up more)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -204,7 +240,10 @@ def tensor_has_hints(t):
 def free_symbols(val: Union[SymInt, torch.Tensor]) -> Set[sympy.Symbol]:
     if isinstance(val, (SymInt, SymFloat)):
-        return val.node.expr.free_symbols
+        if is_symbolic(val):
```
You're right

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (going to write this up more)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (going to write this up more)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented on torch/_subclasses/meta_utils.py:
```diff
@@ -406,9 +406,12 @@ def is_c_of_r(complex_dtype, real_dtype):
+                        sizes,
+                        strides,
+                        storage_offset,
+                    ) = sym_sizes_strides_storage_offset(t, source)
```
TODO: undo this noise

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -20,12 +20,22 @@ def get_tensor_id(tensor, factor):
     _values: torch.Tensor  # type: ignore[assignment]
     _offsets: torch.Tensor
-    _size: Tuple[int, int, int]
+    _size: Tuple[int, torch.SymInt, int]
```
Yeah we'll have to coordinate the land of that one and this one. I don't mind which one lands first. If that one lands first, we'll need to update this one to support generalized jagged yes.

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -83,6 +102,45 @@ def __repr__(self):
         return f"NestedTensor(size={self._size}, offsets={self.offsets}{grad_fn_str})"
 
+    def __tensor_flatten__(self):
+        return ["_values", "_offsets"], (self.requires_grad, self.ragged_size)
```
> also can we just store _size metadata instead of introducing this new ragged_size ?

Good point, getting rid of the duplicate state

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (going to write this up more)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented:
Comments are addressed, this is ready for another round of review!

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

bdhirsh commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -83,6 +112,28 @@ def __repr__(self):
+        offsets = inner_tensors["_offsets"]
+        (requires_grad, ragged_size, *extra_meta) = meta
+
+        if len(extra_meta) > 0:
```
The extra_meta isn't actually optional here, right? Since you always need one to properly reconstruct the subclass

I think I would expect that `__tensor_flatten__()` returns the ragged dimension as another piece of metadata, that way we can plumb it back to this `__tensor_unflatten__()` call when we reconstruct the JaggedTensor.

And then in the fakeifying case, the ragged dimension SymInt is actually the "wrong" SymInt, since it's the one from eager mode, and not the dynamo one we need that actually has a source attached.

bdhirsh replied:
One option (lmk what you think): It might be cleaner not to try to mess with `transform_subclass` at all, and have the manual logic in meta_utils.py be something like:
```
if t.is_nested:
    attrs, ctx = t.__tensor_flatten__()
    transformed_tensors_dict = {}
    for attr in attrs:
        inner_t = getattr(t, attr)
        transformed_tensors_dict[attr] = callback(lambda: empty_create(inner_t, AttrSource(source, attr))
        assert isinstance(ctx, dict) and 'jagged_dim' in ctx
        # Replace the eager jagged dim with our freshly allocated jagged dim that has a source
        ctx['jagged_dim'] =  sym_ragged_size
        return type(t).__tensor_unflatten__(transformed_tensors_dict, ctx)
```
lambda attr, inner_t: callback(
                                lambda: empty_create(
                                    inner_t,
                                    AttrSource(source, attr),
                                )
                            ),

soulitzer replied:
Thanks for the suggestion, this seems like a cleaner option!

(From offline discussion, the ctx is a tuple today, but can be any object, e.g. dict since it is only used by the flatten/unflatten methods and treated as an opaque blob.)

soulitzer replied:
It has been updated now.

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2702,6 +2703,11 @@ def create_symbolic_sizes_strides_storage_offset(
+        if ex.is_nested:
+            # For nested tensor subclasses, we rely on transform subclass logic
+            # to create symbolic sizes/strides/storage_offset.
+            return None, None, None
```
It would seem better to just not call this function at all in this case (and assert `not ex.is_nested`)

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -433,15 +433,49 @@ def empty_create(inner_t, inner_src):
-                                lambda: empty_create(
-                                    inner_t,
-                                    AttrSource(source, attr),
+                        if t.is_nested:
```
@bdhirsh also for you

ezyang commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -44,27 +68,32 @@ def __new__(cls, values, offsets, **kwargs):
+        self._strides = (ragged_size * D, D, 1)
+        self._ragged_idx = 1
+
+        # TODO: error if values requires grad
```
yes please

ezyang reviewed (approved):
please bug Brian for review on the tensor flatten bits. They seem a bit overcomplicated to me but I have trouble evaluating them.

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented:
Updates:
- NestedTensor constructor raises an error if value requires grad
- Skip all value ranges logic for singleton int, e.g. no longer generate 0/1 specialization guards

bdhirsh commented on torch/_subclasses/meta_utils.py:
```diff
@@ -433,15 +436,49 @@ def empty_create(inner_t, inner_src):
+                                        source, TensorProperty.SIZE, 1
+                                    ),
+                                ),
+                                hint=t._size[1],
```
`t._size[1]` is the eager mode ragged size here (holding a SymNode)? Maybe we should assert that?

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -2967,7 +2974,11 @@ def create_symbol(
+            if isinstance(val, int):
+                self.var_to_val[sympy_expr] = sympy.Integer(val)
+            else:
+                self.var_to_val[sympy_expr] = SingletonInt(val.node.singleton_int(), coeff=val.node.singleton_coeff())
```
maybe add a comment that today this codepath only happens for JaggedTensor?

bdhirsh commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3288,22 +3307,28 @@ def hint(s):
                 track_symint(source, t)
                 continue
             assert isinstance(t, Tensorlike)
+            sources_and_tensors = [(source, t)]
```
Double checking my understanding: this used to be set up so that if `t` was a wrapper subclasses, `sources_and_tensors` would only be populated by the inner (plain tensors). Vs. now, `sources_and_tensors` will contain both the `JaggedTensor` subclass, **and** its inner values and offsets tensors (and corresponding source). We need this now because we also need to potentially generate guards on the outer JaggedTensor's ragged dimension. Is that right?

soulitzer replied:
Exactly. Updating the comment to clarify that a bit.

bdhirsh commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -44,27 +68,34 @@ def __new__(cls, values, offsets, **kwargs):
-        # In a later PR, we'll need to accept an additional size argument
-        # to handle dynamic shapes.
-        ragged_dim = get_tensor_id(offsets, coeff=1)
+        if ragged_size is None:
```
Will these two things always happen 1-to-1? (both true or both false)

(1) `ragged_size` is specified in the constructor
(2) `values` and `offsets` are fake tensors.

Maybe we should assert that if it's the case

soulitzer replied:
You're right in that it COULD be that way (as in its only useful if we need to propagate the symbolic ragged size), but that is not the case today just because it seemed easier to me to have the torch dispatch logic, in particular `extract_kwargs` to do the same thing during eager and during tracing: always propagate the ragged_size along if I'm passing the offsets along anyway.

bdhirsh reviewed (approved):
I left a few nits, but overall looks good!

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -i

soulitzer committed:
Update on "Make NestedTensor compilable with eager backend"


In this PR:
- Adds support for strides for jagged tensor (design doc for this coming soon)
- NestedTensor skips automatic dynamic
- Make use of bdhirsh's subclass fakification logic by adding the __tensor_{un,}flatten__ functions.
- Additional logic for fakification: since existing subclass fakification logic does not handle the case where the outer tensor has an additional dimension. We insert one-off logic to (1) insert an extra SingletonSymInt onto the fakified NestedTensor. (2) make sure we call track_symint on both the sizes on the inner and outer tensor during guard creation.

Remaining things that are weird:
- Still need to skip some logic in meta utils for some reason (I was going to write this up more, but decided not to since we're not able to do this anyway for a immediate reason: we cannot arbitrarily compare singleton ints. For now I'm just following Brian's advise from [here](https://github.com/pytorch/pytorch/pull/109171#discussion_r1328137070) )


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot merge -i
